lib/core-integration/src/Test/Integration/Framework/Context.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+
+module Test.Integration.Framework.Context
+    ( Context (..)
+    , PoolGarbageCollectionEvent (..)
+    , TxDescription (..)
+    ) where
+
+import Prelude
+
+import Cardano.CLI
+    ( Port (..) )
+import Cardano.Wallet.Primitive.Types
+    ( EpochNo, NetworkParameters, PoolRetirementCertificate )
+import Cardano.Wallet.Transaction
+    ( DelegationAction )
+import Data.IORef
+    ( IORef )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Text
+    ( Text )
+import GHC.Generics
+    ( Generic )
+import Network.HTTP.Client
+    ( Manager )
+import Numeric.Natural
+    ( Natural )
+import Test.Integration.Faucet
+    ( Faucet )
+
+-- | Context for integration tests.
+--
+data Context t = Context
+    { _cleanup
+        :: IO ()
+        -- ^ A cleanup action.
+    , _manager
+        :: (Text, Manager)
+        -- ^ The underlying base URL and manager used by the wallet client.
+    , _walletPort
+        :: Port "wallet"
+        -- ^ Server TCP port.
+    , _faucet
+        :: Faucet
+        -- ^ Provides access to funded wallets.
+    , _feeEstimator :: TxDescription -> (Natural, Natural)
+        -- ^ A fee estimator.
+    , _networkParameters :: NetworkParameters
+        -- ^ Blockchain parameters for the underlying chain.
+    , _target
+        :: Proxy t
+    , _poolGarbageCollectionEvents
+        :: IORef [PoolGarbageCollectionEvent]
```

Comment:
It feels _a bit_ wrong to me to add such a such a specific thing to the `Context` that is global to all integration tests.

This lead me thinking to: what if we instead added something to the `Context` that allowed individual tests to subscribe whatever subset of the wallet logs they liked? [(Example)](PoolGarbageCollectionEvent) Then it could work for other things than pool garbage collection events, and wouldn't require adding decorators to things.

Reply:
> This lead me thinking to: what if we instead added something to the `Context` that allowed individual tests to subscribe whatever subset of the wallet logs they liked? [(Imagined example)](https://input-output-rnd.slack.com/archives/C4Q7MF25U/p1600699953028500) Then it could work for other things than pool garbage collection events, and wouldn't require adding decorators to things.

@Anviking I do agree that it would be great to have a general mechanism to subscribe to a subset of the wallet logs, instead of this specific mechanism.

However, one issue to solve is that a test might need to access events that occurred **_before_** it started executing.

Specifically, in the case of the `STAKE_POOLS_GARBAGE_COLLECTION_01` test:
- the test needs to examine all historical garbage collection events to check that there was indeed one garbage collection per epoch.
- by the time the test starts, several epochs might already have elapsed.
- if we only start recording events at the start of the test, then we won't be able to access earlier garbage collection events.
- the pool DB decorator is able to solve this problem as it starts recording garbage collection events from the very beginning of time.

Regarding:
> It feels a bit wrong to me to add such a such a specific thing to the Context that is global to all integration tests.

I agree with you that this seems like a very specific addition. However, I think this is likely to be an exceptional case, as it's rare that we need to test something that's not visible through the API.

However, if we do end up wanting to make similar additions in future, I agree it would be worth spend effort making a more general mechanism.

I like the idea of using the logging mechanism for this, as the log messages are already typed, meaning that we can compare values directly, without having to parse values from text.